#--- this script is used to do the go analysis for the result generated by diffbind, you can set the Foldchange cut off
path <- '/WORK/lbyybl/ypj/cross_talk/ChIPseq/Polr1c_degron_Rpb3/diff_peak/'
fold <- log2(2)
file <- 'Ruvbl_deseq2.txt'
name <- 'Plor1c_degron_RPB3'
setwd(path)
suppressMessages(library(ChIPseeker))
suppressMessages(library(org.Mm.eg.db))
suppressMessages(library(clusterProfiler))
suppressMessages(library(TxDb.Mmusculus.UCSC.mm10.knownGene))
suppressMessages(library(ReactomePA))
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
data <- read.table(file,stringsAsFactors = F,row.names = 1,header = T)
up <- data[(data$Fold >= fold & data$FDR<= 0.05),]
down <- data[(data$Fold <= -fold & data$FDR<= 0.05),]

getgeneofpeak <- function(df){
  gr <- GenomicRanges::makeGRangesFromDataFrame(df)
  # gene <- seq2gene(gr, tssRegion = c(-2000, 2000), flankDistance = 3000, TxDb=txdb)
  peakAnno <- annotatePeak(gr, tssRegion=c(-2000, 2000),
                           TxDb=txdb, annoDb="org.Mm.eg.db",#overlap = 'TSS',
                           verbose=FALSE,addFlankGeneInfo=TRUE,
                           flankDistance=5000)
  peakAnno <- as.data.frame(peakAnno)
  
  return(peakAnno)
}

getvalue <- function(v){
  va <- eval(parse(text=v))
  return(va)
}
GO_fun <- function(data_ez,name){
  # browser()
  pathway_data <- enrichPathway(unique(data_ez$geneId),organism = 'mouse',pvalueCutoff = 1,qvalueCutoff = 1)
  # head(pathway_data, 2)
  tmp <- head(pathway_data@result,15)
  rt <- tmp$GeneRatio
  rt <- unlist(lapply(rt, getvalue))
  min <- min(rt)
  max <- max(rt)
  fix <- (max-min)/7
  # pdf(paste0(name,'clust',i,'_enrich.pdf'),width = 9.5,height = 5)
  print(dotplot(pathway_data, showCategory=15)+coord_fixed(fix))
  ggsave(paste0(name,'_Reactome_enrich.pdf'),height = 5,width = 20)
  #print(plot)
  # dev.off()
  pathway_data <- setReadable(pathway_data, OrgDb = org.Mm.eg.db, keyType="ENTREZID")
  fwrite(pathway_data@result,paste0(name,'_ReactomePA.txt'),sep = '\t')
  
  ego_data <- enrichGO(gene         = unique(data_ez$geneId),
                       OrgDb         = org.Mm.eg.db,
                       keyType       = 'ENTREZID',
                       ont           = "BP",
                       pAdjustMethod = "BH",
                       pvalueCutoff  = 1,
                       qvalueCutoff  = 1)
  ego_data <- setReadable(ego_data, OrgDb = org.Mm.eg.db, keyType="ENTREZID")
  tmp <- head(ego_data@result,15)
  rt <- tmp$GeneRatio
  rt <- unlist(lapply(rt, getvalue))
  min <- min(rt)
  max <- max(rt)
  fix <- (max-min)/7
  # pdf(paste0(name,'clust',i,'_GO_BP_enrichment.pdf'),width = 9,height = 5)
  print(dotplot(ego_data, showCategory=15)+coord_fixed(fix))
  ggsave(paste0(name,'_GO_BP_enrichment.pdf'),height = 5,width = 20)
  # dev.off()
  fwrite(ego_data@result,paste0(name,'ego_clust_GO.txt'),sep = '\t')
  
  kegg_data <- enrichKEGG(gene         = unique(data_ez$geneId),
                          organism        = 'mmu',
                          keyType       = 'ncbi-geneid',
                          pAdjustMethod = "BH",
                          pvalueCutoff  = 1,
                          qvalueCutoff  = 1)
  kegg_data <- setReadable(kegg_data, OrgDb = org.Mm.eg.db, keyType="ENTREZID")
  ego_data <- setReadable(ego_data, OrgDb = org.Mm.eg.db, keyType="ENTREZID")
  tmp <- head(kegg_data@result,15)
  rt <- tmp$GeneRatio
  rt <- unlist(lapply(rt, getvalue))
  min <- min(rt)
  max <- max(rt)
  fix <- (max-min)/7
  # pdf(paste0(name,'_kegg_BP',i,'_enrichment.pdf'),width = 9,height = 5)
  print(dotplot(kegg_data, showCategory=15)+coord_fixed(fix))
  ggsave(paste0(name,'_kegg_enrichment.pdf'),height = 5,width = 20)
  # dev.off()
  # pdf(paste0(name,'bar_kegg_BP',i,'_enrichment.pdf'),width = 9,height = 5)
  # print(barplot(kegg_data, showCategory=15))
  # dev.off()
  fwrite(kegg_data@result,paste0(name,'kegg_GO.txt'),sep = '\t')
}

up_gene <- getgeneofpeak(up)
down_gene <- getgeneofpeak(down)
GO_fun(up_gene,paste0(name,'_up'))
GO_fun(down_gene,paste0(name,'_down'))
